File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/ViewportMode.java
package org.maplibre.nativeffi;

/** Viewport orientation mode used by map viewport options. */
public enum ViewportMode {
  DEFAULT(0),
  FLIPPED_Y(1),
  UNKNOWN(-1);

  private final int nativeValue;

  ViewportMode(int nativeValue) {
    this.nativeValue = nativeValue;
  }

  public int nativeValue() {
    if (this == UNKNOWN) {
      throw new IllegalArgumentException("UNKNOWN viewport mode cannot be used as an input");
    }
    return nativeValue;
  }

  public static ViewportMode fromNative(int nativeValue) {
    return switch (nativeValue) {
      case 0 -> DEFAULT;
      case 1 -> FLIPPED_Y;
      default -> UNKNOWN;
    };
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/map/ViewportMode.kt
package org.maplibre.nativeffi.map

/** Viewport orientation mode used by map viewport options. */
public enum class ViewportMode(public val nativeValue: Int) {
  DEFAULT(0),
  FLIPPED_Y(1),
  UNKNOWN(-1);

  public companion object {
    internal fun fromNative(nativeValue: UInt): ViewportMode = fromNative(nativeValue.toInt())

    public fun fromNative(nativeValue: Int): ViewportMode =
      entries.firstOrNull { it.nativeValue == nativeValue } ?: UNKNOWN
  }
}
